crates/pathfinder/src/p2p_network/sync_handlers/v1.rs

```diff
@@ -0,0 +1,579 @@
+    if !classes_for_this_msg.is_empty() {
+        responses.push(block_bodies_response::take_from_class_definitions(
+            block_number,
+            &mut classes_for_this_msg,
+        ));
+    }
+    debug_assert!(classes_for_this_msg.is_empty());
+
+    Ok(())
+}
+
+mod block_bodies_response {
+    use p2p_proto_v1::block::BlockBodyMessage;
+
+    use super::*;
+
+    /// It is assumed that the chunk is not empty
+    pub fn from_class_definition_part(
+        block_number: BlockNumber,
+        class_hash: ClassHash,
+        part: &[u8],
+        parts_count: u32,
+        part_idx: u32,
+    ) -> BlockBodiesResponse {
+        use p2p_proto_v1::state::{Class, Classes};
+        BlockBodiesResponse {
+            block_number: block_number.get(),
+            body_message: BlockBodyMessage::Classes(Classes {
+                domain: 0, // FIXME
+                classes: vec![Class {
+                    compiled_hash: Hash(class_hash.0),
+                    definition: part.to_vec(),
+                    total_parts: Some(parts_count),
+                    part_num: Some(part_idx),
+                }],
+            }),
+        }
+    }
+
+    /// Pops all elements from `class_definitions` leaving the vector empty as if it was just `clear()`-ed,
+    /// so that later on it can be reused.
+    pub fn take_from_class_definitions(
+        block_number: BlockNumber,
+        class_definitions: &mut Vec<(ClassHash, Vec<u8>)>,
+    ) -> BlockBodiesResponse {
+        use p2p_proto_v1::state::{Class, Classes};
+        let classes = class_definitions
+            .drain(..)
+            .map(|(class_hash, definition)| Class {
+                compiled_hash: Hash(class_hash.0),
+                definition,
+                total_parts: None,
+                part_num: None,
+            })
+            .collect();
+        BlockBodiesResponse {
+            block_number: block_number.get(),
+            body_message: BlockBodyMessage::Classes(Classes {
+                domain: 0, // FIXME
+                classes,
+            }),
+        }
+    }
+}
+
+async fn spawn_blocking_get<Request, Response, Getter>(
+    request: Request,
+    storage: &Storage,
+    getter: Getter,
+) -> anyhow::Result<Response>
+where
+    Request: Send + 'static,
+    Response: Send + 'static,
+    Getter: FnOnce(Transaction<'_>, Request) -> anyhow::Result<Response> + Send + 'static,
+{
+    let storage = storage.clone();
+    let span = tracing::Span::current();
+
+    tokio::task::spawn_blocking(move || {
+        let _g = span.enter();
+        let mut connection = storage
+            .connection()
+            .context("Opening database connection")?;
+        let tx = connection
+            .transaction()
+            .context("Creating database transaction")?;
+        getter(tx, request)
+    })
+    .await
+    .context("Database read panic or shutting down")?
+}
+
+async fn send<T>(tx: mpsc::Sender<T>, seq: Vec<T>) -> anyhow::Result<()>
```

Comment:
is there some idiomatic way to do this

Reply:
[PollSender](https://docs.rs/tokio-util/latest/tokio_util/sync/struct.PollSender.html#impl-Sink%3CT%3E-for-PollSender%3CT%3E) is apparently the way to go now if you want to make this a `Sink`, but I think it's fine as-is